### ISSUE TITLE
Modify/Rewrite UDP Reflector for easier client self-deployment

### ISSUE THREAD

**Issue by tlaurion** [inverse-inc/packetfence#1781]
UDP reflector is currently limiting its deployment rate in favor of multiple IP Helper on client infrastructure, even though the later provides PacketFence less precise information when compared to the prior and could be selective at the source and only look and forward interesting packets to PacketFence (DHCPREQUEST and DHCPACK).

IP Helpers listen on layer 3; and are only aware of broadcast packets that also reach their IP addresses. When the packets are directed to a specific IP, those packets are not seen by an IP Helper, but are seen by the DHCP server, where the reflector can be deployed.  This means that the DHCPACK packets are missing from IP Helpers (unicast), while they see DISCOVER, OFFER (broadcast) that PacketFence doesn't really care about. In fact, PacketFence only really cares about DHCPREQUEST and DHCPACK, which can be instructed (currently in a [hackish](https://github.com/inverse-inc/packetfence/issues/1747#issuecomment-253621852) way) to UDP Reflector.

To deploy UDP reflector, the client needs to:
- manually download and install WinPCAP, Microsoft Visual C++ 2010 Redistributable, nssm.
- manually determine which pcap interface matches the system interface he wants to copy the traffic from
- manually create a .bat file and set all program parameters there.
- manually create a new service with nssm with previous create file.

Proposition: 
- UDP Reflector should permit an additional custom BPF field to choose from (it is currently an "UDP" reflector only because it prepends and force "udp and port" in its BPF creation.) 
- An installer should wrap and install dependencies.
  - The installer should propose on which interface to listen.
  - The installer should ask for destination host
  - The installer should propose packet types to look for (BPF filters creation in OR ) and permit custom to add (AND)
  - The installer should install the configuration file and install the service.

**Comment by apkt** [comment]
I totally agree with this request. It makes the whole deployment of PF in out-of-band mode so much easier than using ip-helpers. To achieve this same scenario I have added code into the process_packet subroutine to hack out the stuff that PF is not interested in. It reduces the NODE database table to a much more manageable size and speeds everything up considerably.

**Comment by tlaurion** [comment]
An alpha version is downloadable from [here](https://github.com/inverse-inc/packetfence-dhcp-forwarder).
It was tested on windows 7 and 10, on x64 architecture.

Note that there is currently no eventlog logging. The validation is done from the installer, and refuses to finish the installation if the service wasn't started successfully.

**Comment by tlaurion** [comment]
@apkt look at this [ticket](https://github.com/inverse-inc/packetfence/issues/1747), 
more specifically at this [line](https://github.com/inverse-inc/packetfence/blob/b4e859549f8bea0e2174b545fd9c405b9f178f88/conf/pf.conf.defaults#L66)

Modifying this line to:
`dhcp_filter_by_message_types=DHCPREQUEST,DHCPACK`

instruct PacketFence to only parse those packets through pfdhcplistener.

The goal of the DHCP Forwarder is to actually just forward the interesting packets, filtering them directly at the source, so that PacketFence only receives what it is actually only interested in. All other DHCP types are logged if received and not configured to be neglected in pf.conf.

**Comment by apkt** [comment]
Thanks for the tip but this is not really what I was on about.

Our production DHCP server handles scopes for about 500 different subnets and the UDP reflector
sends all REQUESTS and ACKS from ALL the subnets to the PF server for processing.

The code in lib:pf:dhcp.processor.pm adds the MAC address to the NODES database (if it is not already in there)
irrespective of what subnet scope the ACK/REQUEST is coming from/to.

Consequently the DHCP processor adds the MAC address of every single device on out network to the nodes database
(all PC’s/Servers/VOIP Phones/APs etc.) meaning that our nodes database would contained nearly 200,000+ entries.

Since our PF server is only controlling access to a few subnets it only needs to handle a fraction (<30,000)
of that total.

Without the filtering we have put in place that size of NODE database causes the whole system to grind to a halt as
the various SQL selects take longer and longer to process.

What I would like to have in is the ability to filter the incoming DHCP packets and to ignore any that contain yiaddr and ciaddr
values that the PF server does not have in its configured networks list.

Attached is our modified lib:pf:dhcp:processor.pm which does this but in a very basic (and brutal) way (I am no PERL expert!)

This keeps our nodes database to a manageable 23,000 instead of over 200,000!

Andrew

**Comment by tlaurion** [comment]
@apkt : your modified lib:pf:dhcp:processor.pm was not attached.

Would it fit your need if desired IP ranges would be defined under Configuration->Trapping:"Addresses ranges" so that only those ranges would be taken care of? We could reuse those ranges in DHCP preprocessor so that only IPs defined in those ranges would be considered as nodes under PF.

**Comment by apkt** [comment]
That would be great.

It is mainly to prevent the tables in the database from becoming bloated with every device
on the network.

My mod uses the ‘defined’ routed networks from the interfaces config to filter out the uninteresting DHCP
traffic.

I’ve attached the file again.

The only thing I have had to stop was the injection of the DHCP entry into the OMAPI database as this is
causing the DHCPD process to crash when there are too many hosts entries in the dhcpd.leases files.

Andrew

**Comment by apkt** [comment]
processor.pm - Modded by APKT to stop unwatched (untrapped) DHCP activity populating into NODES table.
[processor.zip](https://github.com/inverse-inc/packetfence/files/661280/processor.zip)

**Comment by julsemaan** [comment]
@tlaurion What is the resolution of this issue ?

**Comment by apkt** [comment]
Funny you should ask that. I was thinking the same thing myself.

Andrew

**Comment by tlaurion** [comment]
This ticket was related to creation of a more user friendly UDP Reflector, which became DHCP Forwarder which is different project: https://github.com/inverse-inc/packetfence-dhcp-forwarder/

@apkt: I tagged you in https://github.com/inverse-inc/packetfence-dhcp-forwarder/issues/10 to see if that proposition would fit your need.

@julsemaan : Resolution of original ticket would be : fixed. I reopened the ticket until clear, so the discussion continue there or here.
